src/libcore/pin.rs

```diff
@@ -191,15 +297,46 @@ impl<P: Deref> Pin<P> {
     /// not guarantee that the data `P` points to is pinned, constructing a
     /// `Pin<P>` is undefined behavior.
     ///
+    /// By using this method, you are making a promise about the `P::Deref` and
+    /// `P::DerefMut` implementations, if they exist. Most importantly, they
+    /// must not move out of their `self` arguments: `Pin::as_mut` and `Pin::as_ref`
+    /// will call `DerefMut::deref_mut` and `Deref::deref` *on the pinned pointer*
+    /// and expect these methods to uphold the pinning invariants.
+    /// Moreover, by calling this method you promise that the reference `P`
+    /// dereferences to will not be moved out of again; in particular, it
+    /// must not be possible to obtain a `&mut P::Target` and then
+    /// move out of that reference (using, for example [`replace`]).
+    ///
+    /// For example, the following is a *violation* of `Pin`'s safety:
+    /// ```
+    /// use std::mem;
+    /// use std::pin::Pin;
+    ///
+    /// fn foo<T>(mut a: T, b: T) {
+    ///     unsafe { let p = Pin::new_unchecked(&mut a); } // should mean `a` can never move again
+    ///     let a2 = mem::replace(&mut a, b);
+    ///     // the address of `a` changed to `a2`'s stack slot, so `a` got moved even
+    ///     // though we have previously pinned it!
+    /// }
+    /// ```
```

Comment:
An example like this one would also be welcome in the module-level docs.

Reply:
What do you think would be a good spot there? They are already quite long.

Reply:
Also I don't think I agree. This example is really specifically about misuse of `new_unchecked`. If that doesn't belong here, nothing does.

Reply:
Agreed. Resolved.